CSS Variables - easier theming and small footprint (Closes #186)

kubawolanin opened the pull request:
Here's my attempt for more accessible theming in HABPanel.
Closes #186 
Fixes #189

* I completely rewrote the `/themes/` part - they are plain CSS files now.
* The variables are visible to the end user on the `:root` level  - there's a posibility to change themes in real time, i.e. change background color when Item has changed and so on :-)
* Footprint of **each theme** decreased from 150-ish kb down to 2 kb! 😮 
* Themes are pretty self explainatory. They could even be stored in openHAB's /config/ or /static/ | /html/ folder so they're more accessible for the 3rd party theme creators?
* All widgets and others now utilize CSS custom properties instead of SCSS variables.
* `vendor.min.js` now contains assets and widgets styles. We might want to rename that in the future?
* My VS Code's CSS Comb extension cleaned up the files a little bit ;-)

As always - I'm open for your remarks :)

Signed-off-by: Kuba Wolanin <hi@kubawolanin.com>

ghys commented:
Hi Kuba,
Just had a quick look at it (stay tuned for something more thorough) and this already looks excellent!

> The variables are visible to the end user on the :root level - there's a posibility to change themes in real time, i.e. change background color when Item has changed and so on :-)

I even had the idea of providing a setting at the dashboard level to override key variables like the background/accent color per dashboard (so they can be more easily told apart), I can definitely see it possible now.

> Footprint of each theme decreased from 150-ish kb down to 2 kb!

🎉 

> Themes are pretty self explainatory. They could even be stored in openHAB's /config/ or /static/ | /html/ folder so they're more accessible for the 3rd party theme creators?

Could be considered, yes. Maybe it's a bit early now. Some documentation of the variables included in the theme, ideally with screenshots, would also be in order.

> `vendor.min.js` now contains assets and widgets styles. We might want to rename that in the future?

Sure, `styles.min.css` or similar.


I also noticed some minor regressions:

![image](https://user-images.githubusercontent.com/2004147/27158648-40c648c2-5168-11e7-99ab-06f03df3364d.png)

Basically it seems to be back to vanilla Bootstrap - the font reverted to Helvetica (and a little bigger), rounded borders have a "retro" feel to them now, some colors have changed to blue-ish, also the color pickers look funky :)
I would be cool to keep everything looking as it is now, the design kind of grew on me :)

But VERY good work overall! Thanks for taking the time 👍

kubawolanin commented:
Hi @ghys! Glad you like it :) 
I've fixed the regression things you've mentioned.
Also renamed `vendor` to `styles` + added two themes already. You can preview [one of them here](https://community.openhab.org/t/new-design-talent-wanted/29718/5?u=kubawolanin).

There's one thing that bothers me in the Knob and Timeline widgets. Their colors are hardcoded in JS which gave me a lot of grief.
Maybe we could create a helper function somewhere to read CSS properties in those places?

```js
        function getStyleProperty(prop) {
            var styles = window.getComputedStyle(document.body);
            return styles.getPropertyValue(prop);
        }
```

and use it like this:
```js
getStyleProperty('--primary-color'); // e.g. "hsl(29, 100%, 50%)" or "rgb(255,255,0)" or "#b00b1e"
```
I'm not sure where's the best place for it so I'll leave it to you ;-)

We might also want to look at intelligent format conversion if needed (i.e. some library requires defining colors in hex format only etc.).

Cheers,
Kuba

ghys commented on web/assets/styles/themes/madras.css:
```diff
@@ -0,0 +1,41 @@
+body,
+.container {
+  background-color: hsl(34, 53%, 82%);
+  background-image: repeating-linear-gradient(45deg, transparent 5px, hsla(197, 62%, 11%, .5) 5px, hsla(197, 62%, 11%, .5) 10px, hsla(5, 53%, 63%, 0) 10px, hsla(5, 53%, 63%, 0) 35px, hsla(5, 53%, 63%, .5) 35px, hsla(5, 53%, 63%, .5) 40px, hsla(197, 62%, 11%, .5) 40px, hsla(197, 62%, 11%, .5) 50px, hsla(197, 62%, 11%, 0) 50px, hsla(197, 62%, 11%, 0) 60px, hsla(5, 53%, 63%, .5) 60px, hsla(5, 53%, 6 ...
```
I just love this stuff ;)
(and other from the url above)

ghys commented on web/assets/styles/themes/madras.css:
```diff
@@ -0,0 +1,41 @@
+  --primary-color: hsl(29, 100%, 50%);
+  --box-bg: rgba(255, 255, 255, .9);
+  --header-bg: var(--body-bg);
+  --widget-design-bg: hsl(0, 0%, 0%);
```
Maybe you want this a bit lighter?
The text is difficult to read:
![image](https://user-images.githubusercontent.com/2004147/27194005-47f82004-5201-11e7-98ff-c2b7c2f47a42.png)

ghys commented on web/assets/styles/themes/material.css:
```diff
@@ -0,0 +1,56 @@
+    .btn-default {
+        color: black;
+    }
+}
```
The syntax above is not supported since this is now vanilla CSS.

ghys commented on web/assets/styles/themes/material.css:
```diff
@@ -0,0 +1,56 @@
+    /* Chart widget */
+    --chart-stroke: #456;
+    --chart-fill: #89a;
+    --chart-tooltip: #123;
```
Could you make this white?
![image](https://user-images.githubusercontent.com/2004147/27194156-d61776c8-5201-11e7-9919-134365588d7d.png)
I know, it's probably not your fault, it was like this before too... 😊

ghys commented on web/assets/styles/themes/material-dark.css:
```diff
@@ -0,0 +1,38 @@
+    --chart-fill: #89a;
+    --chart-tooltip: #123;
+    /* Color picker widget */
+    --colorpicker-border: #3960FF;
```
Could you lowercase this ('ff') as they are lowercase everywhere else?
(in the other themes, too)

ghys commented on web/assets/styles/common.scss:
```diff
@@ -1,217 +1,211 @@
 h2,
 h3.add-widgets {
-  color: $widget-text-color;
+    color: var(--widget-text-color);
```
Maybe it's time to add a variable for this?
It looked OK so far but for not in all cases apparently!
![image](https://user-images.githubusercontent.com/2004147/27194824-625a1fc6-5204-11e7-8188-3031db05b5c3.png)

(Or simply override this in the theme)

kubawolanin replied:
Overrode it [here](https://github.com/kubawolanin/org.openhab.ui.habpanel/blob/c9ac71601a2302fcc975252cbf12ce74b8fb5eb0/web/assets/styles/themes/madras.css#L43).

ghys commented on web/assets/styles/themes/orange-tree.css:
```diff
@@ -0,0 +1,70 @@
+  --widget-text-color: #a4a4a4;
+  --widget-border-radius: 3px;
+  --widget-box-shadow: 0 1px 8px 0 rgba(0, 0, 0, .8), 0 1px 1px 0 rgba(0, 0, 0, .14), 0 2px 1px -1px rgba(0, 0, 0, .12);
+  --icon-color-filter: invert(263%) sepia(100%) hue-rotate(-42deg) saturate(100000%);
```
I think the saturate is a little heavy :)
Looks like this:
![image](https://user-images.githubusercontent.com/2004147/27194559-4ca9e5ae-5203-11e7-9b08-0846838a3355.png)

`invert(1003%) sepia(100%) hue-rotate(-67deg) saturate(10000%)` looks like this:
![image](https://user-images.githubusercontent.com/2004147/27194541-33cdc49c-5203-11e7-907f-37350088fc73.png)

ghys commented:
> Maybe we could create a helper function somewhere to read CSS properties in those places?

@kubawolanin I added this in a filter in the above commit, that way it's easy to use it in templates:

    {{ variable | themeValue:'primary-color') }}

will return `variable` unmodified if it's truthy, or replace it
with the CSS property (`--` is added automatically) if it's falsy.

You can also inject it in controllers or directives:

    ExampleFunc.$inject = ['themeValueFilter'];
    function ExampleFunc(themeValueFilter) {
      var color = themeValueFilter(variable, 'body-bg');
      ...
    }

kubawolanin commented:
@ghys thank you for the review and enhancements. I've applied fixes to the code in c9ac716.
Cheers,
K

ghys commented:
Thanks! Seems good now!

ghys commented:
There are still problems with icon colors when using Firefox (Orange tree theme)...

![image](https://user-images.githubusercontent.com/2004147/27295927-08606f98-551f-11e7-9b1d-f1ddc1775976.png)

ghys commented on web/assets/styles/common.scss:
```diff
@@ -1,6 +1,7 @@
 body {
-    background: var(--body-bg);
+    background: #000;
+    background: var(--body-bg, #000);
```
This sort of trick doesn't work, because of SASS :(
The compiler discards the first line, leaving only the second one which won't be interpreted by older browsers.
Maybe there's a way in SASS to force both lines to appear:
https://stackoverflow.com/questions/44271920/css-variables-with-fallback-for-older-browsers
(have to wrap my head around this)

ghys replied:
https://codepen.io/jakealbaugh/post/css4-variables-with-fallbacks-using-sass

kubawolanin replied:
Hey Yannick,
I suspect it's not because of SASS but [cssmin package](https://www.npmjs.com/package/gulp-cssmin). In the docs it references [clean-css](https://github.com/jakubpawlowicz/clean-css).

We can try to change [this line](https://github.com/kubawolanin/org.openhab.ui.habpanel/blob/366a61df2f779e8e66d0595fa28d31141f654a63/web/gulpfile.js#L50) like so:

```js
        .pipe(sass())
        .pipe(cssmin({
            removeDuplicateRules: false
        }))
        .pipe(rename({
```

And see if that could help.
I don't have my dev environment with me atm, so can't try it myself now.

ghys replied:
Yes it was cssmin, good catch!
I couldn't get gulp-cssmin to understand the options so I replaced it by the (probably better supported) gulp-clean-css.

kubawolanin replied:
Fantastic work! Thanks for pushing it forward man! 🍺

ghys commented:
I don't see any more regressions while testing using a variety of browsers, so lgtm!
Many thanks for your work on this @kubawolanin! 👍
